pennylane/_qubit_device.py

```diff
@@ -0,0 +1,207 @@
+
+            return self._asarray(results)
+
+    def rotate_basis(self, observable):
+        """Rotates the specified wires such that they
+        are in the eigenbasis of the provided observable.
+
+        Args:
+            observable (Observable): the observable that is to be
+                measured
+        """
+        for diag_gate in observable.diagonalizing_gates()[::-1]:
+            self.apply(diag_gate)
+
+    def expval(self, observable):
+        wires = observable.wires
+        par = observable.parameters
+
+        if self.analytic:
+            # exact expectation value
+            eigvals = observable.eigvals
+            prob = np.abs(self._state ** 2)
+            prob = self.marginal_prob(prob, wires=wires)
+            return (eigvals @ prob).real
+
+        # estimate the ev
+        return np.mean(self.sample(observable))
+
+    def var(self, observable):
+        wires = observable.wires
+        par = observable.parameters
+
+        if self.analytic:
+            # exact variance value
+            eigvals = observable.eigvals
+            prob = np.abs(self._state ** 2)
+            prob = self.marginal_prob(prob, wires=wires)
+            return (eigvals ** 2) @ prob - (eigvals @ prob).real ** 2
+
+        return np.var(self.sample(observable))
+
+    def sample(self, observable):
+        if isinstance(observable, qml.Identity):
+            return np.ones([self.shots])
+        self.rotate_basis(observable)
+        # Sampling is supported on software simulators
+        # HW devices will have to provide their own sample() method
+        if isinstance(observable, qml.Hermitian):
+            eigvals = observable.eigvals(np.array(observable.parameters))
+        else:
+            eigvals = observable.eigvals
+
+        eigvals = np.fromiter(eigvals, dtype=np.float64)
+        prob = np.abs(self._state ** 2)
+        prob = self.marginal_prob(prob, wires=observable.wires)
+        return np.random.choice(eigvals, self.shots, p=prob)
+
+    def probability(self, wires=None, values_only=False):
+        """Return the (marginal) probability of each computational basis
+        state from the last run of the device.
+        Args:
+            wires (Sequence[int]): Sequence of wires to return
+                marginal probabilities for. Wires not provided
+                are traced out of the system.
+        Returns:
+            OrderedDict[tuple, float]: Dictionary mapping a tuple representing the state
+            to the resulting probability. The dictionary should be sorted such that the
+            state tuples are in lexicographical order.
+        """
+        if self._state is None:
+            return None
+
+        wires = wires or range(self.num_wires)
+
+        self._probability = np.abs(self._state)**2
+        prob = self.marginal_prob(self._probability, wires)
+        basis_states = itertools.product(range(2), repeat=len(wires))
+        return OrderedDict(zip(basis_states, prob))
```

Comment:
Note that, if an `self._sample` attribute is added as per my previous comment, this should also generalize to hardware devices!
```python
if hardware:
    # count the number of occurrences in `self._sample`, and divide by self.shots
```

Perhaps we should have a `self._hardware` flag for hardware devices to automatically switch to this logic?